federation/pkg/federation-controller/replicaset/replicasetcontroller.go

```diff
@@ -0,0 +1,448 @@
+	}
+}
+
+func (frsc *ReplicaSetController) deliverFedReplicaSetObj(obj interface{}, delay time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	frsc.deliverReplicaSetByKey(key, delay, false)
+}
+
+func (frsc *ReplicaSetController) deliverReplicaSetByKey(key string, delay time.Duration, failed bool) {
+	if failed {
+		frsc.replicaSetBackoff.Next(key, time.Now())
+		delay = delay + frsc.replicaSetBackoff.Get(key)
+	} else {
+		frsc.replicaSetBackoff.Reset(key)
+	}
+	frsc.replicasetDeliverer.DeliverAfter(key, nil, delay)
+}
+
+func (frsc *ReplicaSetController) worker() {
+	for {
+		item, quit := frsc.replicasetWorkQueue.Get()
+		if quit {
+			return
+		}
+		key := item.(string)
+		err := frsc.reconcileReplicaSet(key)
+		frsc.replicasetWorkQueue.Done(item)
+		if err != nil {
+			glog.Errorf("Error syncing cluster controller: %v", err)
+			frsc.deliverReplicaSetByKey(key, 0, true)
+		}
+	}
+}
+
+func (frsc *ReplicaSetController) schedule(frs *extensionsv1.ReplicaSet, clusters []*fedv1.Cluster,
+	expected map[string]int64, actual map[string]int64) map[string]int64 {
+	// TODO: integrate real scheduler
+
+	plnr := frsc.defaultPlanner
+	frsPref, err := parseFederationReplicaSetReference(frs)
+	if err != nil {
+		glog.Info("Invalid ReplicaSet specific preference, use default. rs: %v, err: %v", frs, err)
+	}
+	if frsPref != nil { // create a new planner if user specified a preference
+		plnr = planner.NewPlanner(frsPref)
+	}
+
+	replicas := int64(*frs.Spec.Replicas)
+	var clusterNames []string
+	for _, cluster := range clusters {
+		clusterNames = append(clusterNames, cluster.Name)
+	}
+	scheduleResult, _ := plnr.Plan(replicas, clusterNames, expected, actual)
+	// make sure the return contains clusters need to zero the replicas
+	result := make(map[string]int64)
+	for clusterName := range expected {
+		result[clusterName] = 0
+	}
+	for clusterName, replicas := range scheduleResult {
+		result[clusterName] = replicas
+	}
+	return result
+}
+
+func (frsc *ReplicaSetController) reconcileReplicaSet(key string) error {
+	if !frsc.isSynced() {
+		frsc.deliverReplicaSetByKey(key, clusterAvailableDelay, false)
+		return nil
+	}
+
+	glog.Infof("Start reconcile replicaset %q", key)
+	startTime := time.Now()
+	defer glog.Infof("Finished reconcile replicaset %q (%v)", key, time.Now().Sub(startTime))
+
+	obj, exists, err := frsc.replicaSetStore.Store.GetByKey(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		// don't delete local replicasets for now
+		return nil
+	}
+	frs := obj.(*extensionsv1.ReplicaSet)
+
+	clusters, err := frsc.fedReplicaSetInformer.GetReadyClusters()
+	if err != nil {
+		return err
+	}
+
+	// collect current status and do schedule
+	allPods, err := frsc.fedPodInformer.GetTargetStore().List()
+	if err != nil {
+		return err
+	}
+	podStatus, err := AnalysePods(frs, allPods, time.Now())
+	expected := make(map[string]int64)
```

Comment:
in theory it could be running.
however the reconciliation often happens before the pod status became stable which I'd rather to think that all scheduled (expected) pods are running which would help prevent rebalancing but not keep the stucked pods since we have the capacity as a second parameter.